Fix native_batch_norm_backward returning non-channels_last_3d grad

mikaylagawarecki opened the pull request:
Fix #107199

Checked out https://github.com/pytorch/pytorch/pull/106104 which caught this locally and verified that https://github.com/pytorch/pytorch/blob/551124f67090b7c672e9833ed10f41d31c57747c/torch/testing/_internal/common_modules.py#L2635-L2642 with the `p['device'] == 'cuda'` part shifted to `device_type = 'cuda'` now succeeds

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107270

mikaylagawarecki committed:
Fix native_batch_norm_backward returning non-channels_last_3d_grad

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix native_batch_norm_backward returning non-channels_last_3d grad"

Fix #107199




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -211,7 +211,7 @@ Tensor batch_norm_elementwise_backward_train(
         at::scalar_tensor(1.0, input.options().dtype(mean.scalar_type()));
 
-    Tensor grad_input = at::empty(input.sizes(), grad_out.options());
+    Tensor grad_input = at::empty(input.sizes(), grad_out.options()).to(input.suggest_memory_format());
```
add this to the options() argument of empty. No need for a full copy!

mikaylagawarecki replied:
my bad, fixed!

mikaylagawarecki committed:
Update on "Fix native_batch_norm_backward returning non-channels_last_3d grad"

Fix #107199

Checked out https://github.com/pytorch/pytorch/pull/106104 which caught this locally and verified that https://github.com/pytorch/pytorch/blob/551124f67090b7c672e9833ed10f41d31c57747c/torch/testing/_internal/common_modules.py#L2635-L2642 now succeeds




[ghstack-poisoned]

albanD reviewed (approved):
SGTM

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki commented:
@pytorchbot merge -f "linux-focal-rocm5.6-py3.8 / test (default, 1, 3, linux.rocm.gpu) test_foreach failure is unrelated"
